### DISCUSSION TITLE
cluster-operator doesn't create external listener

### DISCUSSION THREAD

**Discussion by geniymaxdem** [strimzi/strimzi-kafka-operator#6286] [General]
HI! I need your help
I use cluster-operator to create a kafka-cluster
My platform - k8s version v1.22.3
When I apply yaml file to configure kafka-cluster, cluster-operator don't create a external listener
`apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: kafka-cluster
  namespace: kafka
spec:
  kafka:
    version: 3.0.0
    replicas: 3
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
      - name: external
        port: 9094
        type: ingress
        tls: true
        configuration:
          brokerCertChainAndKey:
            secretName: kafka-secret-tls
            certificate: tls.crt
            key: tls.key
          class: nginx
          bootstrap:
            host: kafka.service.rowi.tech
          brokers:
          - broker: 0
            host: kafka-0.service.rowi.tech
          - broker: 1
            host: kafka-1.service.rowi.tech
          - broker: 2
            host: kafka-2.service.rowi.tech
............
    config:
      # start manual config
      ssl.cipher.suites: "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
      ssl.enabled.protocols: "TLSv1.2"
      ssl.protocol: "TLSv1.2"
      log.retention.hours: 24
      log.segment.bytes: 104857600
      log.retention.bytes: 1073741824
      log.retention.check.interval.ms: 300000
      log.cleaner.enable: true
      `
status of  kafka-cluster
`
kubectl -n kafka get kafka kafka-cluster -o yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  creationTimestamp: "2022-02-01T10:54:01Z"
  generation: 2
  name: kafka-cluster
  namespace: kafka
  resourceVersion: "27797783"
  uid: b2c4e679-d534-403e-8fac-be519f4aa35d
spec:
  cruiseControl:
    metricsConfig:
      type: jmxPrometheusExporter
      valueFrom:
        configMapKeyRef:
          key: cruise-metrics-config.yml
          name: cruise-control-metrics
  entityOperator:
    topicOperator: {}
    userOperator: {}
  kafka:
    config:
      default.replication.factor: 3
      inter.broker.protocol.version: "3.0"
      log.cleaner.enable: true
      log.cleanup.policy: delete
      log.retention.bytes: 1073741824
      log.retention.check.interval.ms: 300000
      log.retention.hours: 24
      log.segment.bytes: 104857600
      min.insync.replicas: 2
      offsets.topic.replication.factor: 3
      ssl.cipher.suites: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
      ssl.enabled.protocols: TLSv1.2
      ssl.protocol: TLSv1.2
      transaction.state.log.min.isr: 2
      transaction.state.log.replication.factor: 3
    listeners:
    - name: plain
      port: 9092
      tls: false
      type: internal
    - name: tls
      port: 9093
      tls: true
      type: internal
    - configuration:
        bootstrap:
          host: kafka.service.rowi.tech
        brokerCertChainAndKey:
          certificate: tls.crt
          key: tls.key
          secretName: kafka-secret-tls
        brokers:
        - broker: 0
          host: kafka-0.service.rowi.tech
        - broker: 1
          host: kafka-1.service.rowi.tech
        - broker: 2
          host: kafka-2.service.rowi.tech
        class: nginx
      name: external
      port: 9094
      tls: true
      type: ingress
    metricsConfig:
      type: jmxPrometheusExporter
      valueFrom:
        configMapKeyRef:
          key: kafka-metrics-config.yml
          name: kafka-metrics
    replicas: 3
    storage:
      type: jbod
      volumes:
      - deleteClaim: false
        id: 0
        size: 45Gi
        type: persistent-claim
    version: 3.0.0
  kafkaExporter:
    enableSaramaLogging: true
    groupRegex: .*
    logging: debug
    resources:
      limits:
        cpu: 500m
        memory: 128Mi
      requests:
        cpu: 200m
        memory: 64Mi
    topicRegex: .*
  zookeeper:
    metricsConfig:
      type: jmxPrometheusExporter
      valueFrom:
        configMapKeyRef:
          key: zookeeper-metrics-config.yml
          name: zookeeper-metrics-config
    replicas: 3
    storage:
      deleteClaim: false
      size: 15Gi
      type: persistent-claim
status:
  conditions:
  - lastTransitionTime: "2022-02-01T11:21:07.064Z"
    message: Exceeded timeout of 300000ms while waiting for Ingress resource kafka-cluster-kafka-bootstrap
      in namespace kafka to be addressable
    reason: TimeoutException
    status: "True"
    type: NotReady
  listeners:
  - addresses:
    - host: kafka-cluster-kafka-bootstrap.kafka.svc
      port: 9092
    bootstrapServers: kafka-cluster-kafka-bootstrap.kafka.svc:9092
    type: plain
  - addresses:
    - host: kafka-cluster-kafka-bootstrap.kafka.svc
      port: 9093
    bootstrapServers: kafka-cluster-kafka-bootstrap.kafka.svc:9093
    type: tls
  observedGeneration: 2
`

As you can see, cluster-operator don't create a external listener with port 9094
How can i resolve my problem?
Thanks a lot.

**Comment by scholzj** [comment]
For a start, you would probably need to format the YAML as a code to make it readable.

**Reply by geniymaxdem** [reply]
File in attach

**Comment by geniymaxdem** [comment]
[kafka-cluster-ext.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7978261/kafka-cluster-ext.txt)

**Comment by geniymaxdem** [comment]
[cluster-status.txt](https://github.com/strimzi/strimzi-kafka-operator/files/7978282/cluster-status.txt)

**Reply by scholzj** [reply]
Well, it tells you the error, or?

```
  - lastTransitionTime: "2022-02-01T11:21:07.064Z"
    message: Exceeded timeout of 300000ms while waiting for Ingress resource kafka-cluster-kafka-bootstrap
      in namespace kafka to be addressable
    reason: TimeoutException
    status: "True"
    type: NotReady
```

It looks like your Ingress does not work and does not set the address on the Ingress resources in their status section.

**Comment by geniymaxdem** [comment]
all endpoints present
```
kubectl -n kafka get endpoints
NAME                                     ENDPOINTS                                                             AGE
kafka-cluster-cruise-control             10.233.96.68:9090                                                     75m
kafka-cluster-kafka-0                    10.233.98.228:9094                                                    54m
kafka-cluster-kafka-1                    10.233.98.229:9094                                                    54m
kafka-cluster-kafka-2                    10.233.96.94:9094                                                     54m
kafka-cluster-kafka-bootstrap            10.233.96.94:9093,10.233.98.228:9093,10.233.98.229:9093 + 6 more...   76m
kafka-cluster-kafka-brokers              10.233.96.94:9093,10.233.98.228:9093,10.233.98.229:9093 + 9 more...   76m
kafka-cluster-kafka-external-bootstrap   10.233.96.94:9094,10.233.98.228:9094,10.233.98.229:9094               54m
kafka-cluster-zookeeper-client           10.233.96.47:2181,10.233.98.226:2181,10.233.98.227:2181               77m
kafka-cluster-zookeeper-nodes            10.233.96.47:2888,10.233.98.226:2888,10.233.98.227:2888 + 6 more...   77m
```

**Reply by geniymaxdem** [reply]
This all ingress fir cluster-kafka
```
kubectl -n kafka get ing
NAME                            CLASS   HOSTS                       ADDRESS   PORTS     AGE
kafka-cluster-kafka-0           nginx   kafka-0.service.rowi.tech             80, 443   55m
kafka-cluster-kafka-1           nginx   kafka-1.service.rowi.tech             80, 443   55m
kafka-cluster-kafka-2           nginx   kafka-2.service.rowi.tech             80, 443   55m
kafka-cluster-kafka-bootstrap   nginx   kafka.service.rowi.tech               80, 443   55m
```
All svc in cluster
```
kubectl -n kafka get svc
NAME                                     TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                               AGE
kafka-cluster-cruise-control             ClusterIP   10.233.24.28    <none>        9090/TCP                              77m
kafka-cluster-kafka-0                    ClusterIP   10.233.9.192    <none>        9094/TCP                              56m
kafka-cluster-kafka-1                    ClusterIP   10.233.9.108    <none>        9094/TCP                              56m
kafka-cluster-kafka-2                    ClusterIP   10.233.25.211   <none>        9094/TCP                              56m
kafka-cluster-kafka-bootstrap            ClusterIP   10.233.29.185   <none>        9091/TCP,9092/TCP,9093/TCP            78m
kafka-cluster-kafka-brokers              ClusterIP   None            <none>        9090/TCP,9091/TCP,9092/TCP,9093/TCP   78m
kafka-cluster-kafka-external-bootstrap   ClusterIP   10.233.19.55    <none>        9094/TCP                              56m
kafka-cluster-zookeeper-client           ClusterIP   10.233.56.13    <none>        2181/TCP                              79m
kafka-cluster-zookeeper-nodes            ClusterIP   None            <none>        2181/TCP,2888/TCP,3888/TCP            79m
```
but, I don't see external listener in cluster
Why?
Port 9094 not presented in kafka-broker
```
kubectl -n kafka exec -it kafka-cluster-kafka-0 /bin/sh
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
sh-4.4$ netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp6       0      0 :::9404                 :::*                    LISTEN      65/java             
tcp6       0      0 :::34191                :::*                    LISTEN      65/java             
tcp6       0      0 :::9093                 :::*                    LISTEN      65/java             
tcp6       0      0 :::9092                 :::*                    LISTEN      65/java             
tcp6       0      0 :::9091                 :::*                    LISTEN      65/java             
tcp6       0      0 :::9090                 :::*                    LISTEN      65/java 
```

**Reply by scholzj** [reply]
As I said above, you are missing the `address` in the Ingress status section. You can see it even in your listing of the ingresses above. That suggests that your Ingress controller didn't accepted it for some reason.

**Reply by geniymaxdem** [reply]
It's normal status for my ingress controller
Can you explaine, why pod of broker don't have aport 9094 ?
```
kubectl -n kafka describe po kafka-cluster-kafka-1
Name:         kafka-cluster-kafka-1
Namespace:    kafka
Priority:     0
Node:         master2/10.46.25.12
Start Time:   Tue, 01 Feb 2022 13:54:27 +0300
Labels:       app.kubernetes.io/instance=kafka-cluster
              app.kubernetes.io/managed-by=strimzi-cluster-operator
              app.kubernetes.io/name=kafka
              app.kubernetes.io/part-of=strimzi-kafka-cluster
              controller-revision-hash=kafka-cluster-kafka-6cb97df8d7
              statefulset.kubernetes.io/pod-name=kafka-cluster-kafka-1
              strimzi.io/cluster=kafka-cluster
              strimzi.io/kind=Kafka
              strimzi.io/name=kafka-cluster-kafka
Annotations:  cni.projectcalico.org/containerID: 7c7c8d68d9a9b95ee57b133fc5d9ad4872e9f4f30a14e34f06eee809719d0623
              cni.projectcalico.org/podIP: 10.233.98.229/32
              cni.projectcalico.org/podIPs: 10.233.98.229/32
              strimzi.io/broker-configuration-hash: 45d76058b9a5a7ffe371fed61cc330dce576285b0f4f16bff874a88f03db4306de59264241488d1b
              strimzi.io/clients-ca-cert-generation: 0
              strimzi.io/cluster-ca-cert-generation: 0
              strimzi.io/generation: 0
              strimzi.io/inter-broker-protocol-version: 3.0
              strimzi.io/kafka-version: 3.0.0
              strimzi.io/log-message-format-version: 3.0
              strimzi.io/logging-appenders-hash: e893ac9f32f7958042d1c82477e3484b0f8ebad5
              strimzi.io/storage: {"volumes":[{"type":"persistent-claim","size":"45Gi","id":0}],"type":"jbod"}
Status:       Running
IP:           10.233.98.229
IPs:
  IP:           10.233.98.229
Controlled By:  StatefulSet/kafka-cluster-kafka
Containers:
  kafka:
    Container ID:  cri-o://d4960f9903ab09aa43affbb825e67030aaa614680deb5f9112851816fb5ff374
    Image:         quay.io/strimzi/kafka:0.27.0-kafka-3.0.0
    Image ID:      quay.io/strimzi/kafka@sha256:1886d49d070ac81aa3a608190fc1ce54e4b315e7a2d87dcac7156f520e44e961
    Ports:         9090/TCP, 9091/TCP, 9092/TCP, 9093/TCP, 9404/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP, 0/TCP, 0/TCP
    Command:
      /opt/kafka/kafka_run.sh
    State:          Running
      Started:      Tue, 01 Feb 2022 13:54:30 +0300
    Ready:          True
    Restart Count:  0
    Liveness:       exec [/opt/kafka/kafka_liveness.sh] delay=15s timeout=5s period=10s #success=1 #failure=3
    Readiness:      exec [test -f /var/opt/kafka/kafka-ready] delay=15s timeout=5s period=10s #success=1 #failure=3
    Environment:
      KAFKA_METRICS_ENABLED:         true
      STRIMZI_KAFKA_GC_LOG_ENABLED:  false
      KAFKA_HEAP_OPTS:               -Xms128M
    Mounts:
      /opt/kafka/broker-certs from broker-certs (rw)
      /opt/kafka/client-ca-certs from client-ca-cert (rw)
      /opt/kafka/cluster-ca-certs from cluster-ca (rw)
      /opt/kafka/custom-config/ from kafka-metrics-and-logging (rw)
      /tmp from strimzi-tmp (rw)
      /var/lib/kafka/data-0 from data-0 (rw)
      /var/opt/kafka from ready-files (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-wrw4b (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  data-0:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-0-kafka-cluster-kafka-1
    ReadOnly:   false
  strimzi-tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  1Mi
  cluster-ca:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kafka-cluster-cluster-ca-cert
    Optional:    false
  broker-certs:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kafka-cluster-kafka-brokers
    Optional:    false
  client-ca-cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kafka-cluster-clients-ca-cert
    Optional:    false
  kafka-metrics-and-logging:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      kafka-cluster-kafka-config
    Optional:  false
  ready-files:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  1Ki
  kube-api-access-wrw4b:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
```

**Reply by scholzj** [reply]
That it will not work with Strimzi. Strimzi expects it to set the address in the status. This is what the Kubernetes Nginx Ingress controller which we support and test with does. Because that failed, the operator does not know how to configure the broker and did not add the port to the pods.

**Reply by geniymaxdem** [reply]
> As I said above, you are missing the `address` in the Ingress status section. You can see it even in your listing of the ingresses above. That suggests that your Ingress controller didn't accepted it for some reason.

Thanks a lot for help

I find solution - I insert some params in config-map for ingress-controller
external-status-address: 10.46.17.109
This is IP of my load-balancer
```
apiVersion: v1
data:
client-max-body-size: 20m
external-status-address: 10.46.17.109
proxy-connect-timeout: 10s
proxy-read-timeout: 10s
```
After that - kafka-cluster created correctly

**Reply by scholzj** [reply]
Interesting. I didn't know about such configuration. Seems to be set automatically in all my environments. But glad you made it work 👍

**Comment by geniymaxdem** [comment]
describe ingress
```
kubectl -n kafka describe ing kafka-cluster-kafka-bootstrap
Name:             kafka-cluster-kafka-bootstrap
Namespace:        kafka
Address:          
Default backend:  default-http-backend:80 (<error: endpoints "default-http-backend" not found>)
TLS:
  SNI routes kafka.service.rowi.tech
Rules:
  Host                     Path  Backends
  ----                     ----  --------
  kafka.service.rowi.tech  
                           /   kafka-cluster-kafka-external-bootstrap:9094 (10.233.96.94:9094,10.233.98.228:9094,10.233.98.229:9094)
Annotations:               ingress.kubernetes.io/ssl-passthrough: true
                           nginx.ingress.kubernetes.io/backend-protocol: HTTPS
                           nginx.ingress.kubernetes.io/ssl-passthrough: true
Events:                    <none>
```
